addons/mass_mailing/controllers/main.py

```diff
@@ -118,8 +118,7 @@ def mailing_unsubscribe_placeholder_link(self, **post):
     @http.route('/mail/track/<int:mail_id>/<string:token>/blank.gif', type='http', auth='public')
     def track_mail_open(self, mail_id, token, **post):
         """ Email tracking. """
-        mail = request.env['mail.mail'].sudo().browse(mail_id).exists()
-        if not mail or not consteq(token, mail._generate_mail_recipient_token()):
+        if not consteq(token, tools.hmac(request.env(su=True), 'mass_mailing-mail_mail-open', mail_id)):
```

Comment:
Issue is that if one modifies the method, this one will break easily ... ok there are some tests but that's not consistent. Purpose of the cleaning was to move the generation into a single method.

Could we make '_generate_mail_recipient_token' accept an optional ID (in stable) and make it model + using a mail_id in master ?